File: Competitive Programming/Blind 75 Must Do Leetcode/Climbing Stairs.cpp
/*
https://leetcode.com/problems/climbing-stairs/
70. Climbing Stairs
Easy

8813

261

Add to List

Share
You are climbing a staircase. It takes n steps to reach the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

 

Example 1:

Input: n = 2
Output: 2
Explanation: There are two ways to climb to the top.
1. 1 step + 1 step
2. 2 steps
Example 2:

Input: n = 3
Output: 3
Explanation: There are three ways to climb to the top.
1. 1 step + 1 step + 1 step
2. 1 step + 2 steps
3. 2 steps + 1 step
 

Constraints:

1 <= n <= 45
*/

//DP
/**
Complexity Analysis
Time complexity : O(n)O(n). Single loop upto nn.
Space complexity : O(n)O(n). dpdp array of size nn is used. 
**/
//Runtime: 4 ms, faster than 100.00% of C++ online submissions for Climbing Stairs.
//Memory Usage: 8.5 MB, less than 55.79% of C++ online submissions for Climbing Stairs.
class Solution
{
public:
    int climbStairs(int n)
    {
        if (n == 1)
            return 1;
        vector<int> steps(n);

        // cout << "size: " << steps.size() << endl;

        steps[0] = 1; //1+0 step from the top
        steps[1] = 2; //1+1 step from the top

        for (int i = 2; i < n; i++)
        {
            steps[i] = steps[i - 1] + steps[i - 2];
            // cout << steps[i] << endl;
        }

        return steps[n - 1];
    }
};

//Fibonacci number
/**
Complexity Analysis
Time complexity : O(n)O(n). Single loop upto nn is required to calculate n^{th}n 
th fibonacci number.
Space complexity : O(1)O(1). Constant space is used. 
**/

//Runtime: 4 ms, faster than 100.00% of C++ online submissions for Climbing Stairs.
//Memory Usage: 8.2 MB, less than 97.68% of C++ online submissions for Climbing Stairs.
class Solution
{
public:
    int climbStairs(int n)
    {
        if (n <= 2)
            return n;

        int last = 1, last2 = 2;

        for (int i = 2; i < n; i++)
        {
            int tmp = last2;
            last2 = last + last2;
            last = tmp;
            // cout << last << " " << last2 << endl;
        }
        // cout << endl;

        return last2;
    }
};

//DP and Fibonacci, no edge case
//Runtime: 0 ms, faster than 100.00% of C++ online submissions for Climbing Stairs.
//Memory Usage: 5.9 MB, less than 100.00% of C++ online submissions for Climbing Stairs.
class Solution
{
public:
    int climbStairs(int n)
    {
        // vector<int> dp(n+1, 0);

        //0: padding, used when dp[2] = dp[1] + dp[0]
        // dp[0] = dp[1] = 1;
        int last1 = 1, last2 = 1;

        for (int i = 2; i <= n; i++)
        {
            // dp[i] = dp[i-1] + dp[i-2];
            last2 += last1;
            swap(last1, last2);
        }

        // return dp[n];
        return last1;
    }
};

/**
Approach 5: Binets Method
**/

/**
Complexity Analysis
Time complexity : O(\log n)O(logn). Traversing on \log nlogn bits.
Space complexity : O(1)O(1). Constant space is used.
**/

/**
class Solution {
public:
    void multiply(const vector<vector<long long>>& a, const vector<vector<long long>>& b, vector<vector<long long>>& c){
        for(int i = 0; i < 2; i++){
            for(int j = 0; j < 2; j++){
                c[i][j] = a[i][0] * b[0][j] + a[i][1] * b[1][j];
            }
        }
    }
    
    void pow(vector<vector<long long>>& a, int n, vector<vector<long long>>& r){
        vector<vector<long long>> tmp = {{0,0}, {0,0}};
        while(n > 0){
            //if(n%2 == 1){
            if((n&1) == 1){
                multiply(r, a, tmp);
                r = tmp;
            }
            //n/=2;
            n >>= 1;
            multiply(a, a, tmp);
            a = tmp;
        }
    }
    
    int climbStairs(int n) {
        vector<vector<long long>> q = {{1,1}, {1,0}};
        vector<vector<long long>> r = {{1,0}, {0,1}};
        pow(q, n, r);
        return r[0][0];
    }
};
**/

/**
Approach 6: Fibonacci Formula
**/

/**
Complexity Analysis
Time complexity : O(\log n)O(logn). powpow method takes \log nlogn time.
Space complexity : O(1)O(1). Constant space is used.
**/

/**
class Solution {
public:
    int climbStairs(int n) {
        double sqrt5 = sqrt(5);
        double fibn = pow((1+sqrt5)/2, n+1) - pow((1-sqrt5)/2, n+1);
        return (int)(fibn/sqrt5);
    }
};
**/

File: competitive-programming/Blind 75 Must Do Leetcode/Climbing Stairs.py
'''https://leetcode.com/problems/climbing-stairs/
70. Climbing Stairs
Easy

8813

261

Add to List

Share
You are climbing a staircase. It takes n steps to reach the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

 

Example 1:

Input: n = 2
Output: 2
Explanation: There are two ways to climb to the top.
1. 1 step + 1 step
2. 2 steps
Example 2:

Input: n = 3
Output: 3
Explanation: There are three ways to climb to the top.
1. 1 step + 1 step + 1 step
2. 1 step + 2 steps
3. 2 steps + 1 step
 

Constraints:

1 <= n <= 45'''

# for given n steps stairs
# Each time you can either climb 1 or 2 steps.
# In how many distinct ways can you climb to the top?


class Solution:
    def climbStairs(self, n: int) -> int:
        """
        Find the number of distinct ways to climb to the top of the staircase.

        Args:
        n: int, the number of steps to reach the top.

        Returns:
        int, the number of distinct ways to climb to the top.
        """
        if n == 1:
            return 1
        elif n == 2:
            return 2

        # Initialize variables to store the number of ways to reach the current and previous steps
        prev_step = 1
        curr_step = 2

        # Iterate from step 3 to n, calculating the number of ways to reach each step
        for _ in range(3, n + 1):
            # Calculate the number of ways to reach the current step as the sum of ways to reach the previous two steps
            next_step = prev_step + curr_step
            # Update the values of prev_step and curr_step for the next iteration
            prev_step, curr_step = curr_step, next_step

        return curr_step


def brute_force(n):
    if n == 1 or n == 2:
        return n
    return brute_force(n-1)+brute_force(n-2)


memo = {}


def recursion_memo(n, memo):

    if n == 1 or n == 2:
        return n
    if n in memo.keys():
        return memo[n]
    value = recursion_memo(n-1, memo) + recursion_memo(n-2, memo)
    memo.update({n: value})
    return memo[n]


def dynamic(n):
    if n == 1 or n == 2:
        return n
    result = [0]*(n)
    result[0] = 1
    result[1] = 2
    for i in range(2, n):
        result[i] = result[i-1]+result[i-2]
    return result[n-1]


if __name__ == '__main__':
    print(recursion_memo(10, memo))
